pkg/activator/endpointlistener.go

```diff
@@ -0,0 +1,100 @@
+package activator
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/knative/serving/pkg/apis/serving"
+	pkghttp "github.com/knative/serving/pkg/http"
+	"go.uber.org/zap"
+)
+
+// Endpoint is a fully-qualified domain name / port pair for an active revision.
+type Endpoint struct {
+	FQDN string
+	Port int32
+}
+
+type EndpointListener struct {
+	mux     sync.Mutex
+	Logger  *zap.SugaredLogger
+	watches map[string]*endpointWatch
+}
+
+type endpointWatch struct {
+	el        *EndpointListener
+	bcastCh   chan struct{}
+	namespace string
+	revision  string
+	refs      uint
+	endpoint  *Endpoint
+}
+
+func flattenNamespaceRevision(namespace, revision string) string {
+	return fmt.Sprintf("%s/%s", namespace, revision)
+}
+
+func NewEndpointListener(logger *zap.SugaredLogger) *EndpointListener {
+	return &EndpointListener{
+		Logger:  logger,
+		watches: make(map[string]*endpointWatch),
+	}
+}
+
+func (el *EndpointListener) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	namespace := pkghttp.LastHeaderValue(r.Header, serving.ActivatorRevisionHeaderNamespace)
+	name := pkghttp.LastHeaderValue(r.Header, serving.ActivatorRevisionHeaderName)
+	nsr := flattenNamespaceRevision(namespace, name)
+
+	el.mux.Lock()
+	defer el.mux.Unlock()
+
+	if watch, ok := el.watches[nsr]; ok {
+		watch.endpoint = &Endpoint{}
+		close(watch.bcastCh)
+	}
+}
+
+// TODO(greghaynes) We should add a timeout equal to our activator retry delay for the lifetime of an entire EndpointWatch
+func (el *EndpointListener) WaitForEndpoint(namespace, revision string, closeCh <-chan struct{}) *Endpoint {
+	nsr := flattenNamespaceRevision(namespace, revision)
+
+	el.mux.Lock()
```

Comment:
this feels unsafe.
Can you pull into a func and do lock/defer unlock pair there.